#764 - Introduce Sliding Cache to Constant Expression Helper

TWhidden opened the pull request:
Introduce a Sliding Cache to ConstantExpressionHelper.  
- Use of DateTime (UTC) instead of Stopwatch for expire period. lightweight
- Should still work with older frameworks - Still using same backing store, just with a private Tuple that holds the last accessed time
- Introduced new internal type `ThreadSafeSlidingCache` which has a constructor option of TTL for the object expire period, as well as an override for the cleanup frequency - default of 10 seconds atm - figured that would not add a huge strain, but can be adjusted as desired. did not know where we would put this class in the folder structure, feel free to refactor. It uses a "Lazy" Expire, so there is no timer, but objects expire when the cache is accessed. 
- TTL default for objects built in Release is sliding cache value of 10 minutes. This is just an arbitrary number that seems reasonable. 
- Test Case created - Passes on my end - feel free to refactor the name of the test

StefH commented on src/System.Linq.Dynamic.Core/Parser/ThreadSafeSlidingCache.cs:
```diff
@@ -0,0 +1,80 @@
+
+namespace System.Linq.Dynamic.Core.Parser
+{
+    internal class ThreadSafeSlidingCache<T1, T2> where T1 : notnull where T2 : notnull
```
Rename to TKey

TWhidden replied:
Will do the same for `TValue`

StefH reviewed (changes requested):
Thanks for this PR.

Can you look at the review comments please?

TWhidden commented on src/System.Linq.Dynamic.Core/Parser/ThreadSafeSlidingCache.cs:
```diff
@@ -0,0 +1,80 @@
+        {
+            _cache = new ConcurrentDictionary<T1, (T2, DateTime)>();
+            _timeToLive = timeToLive;
+            _cleanupFrequency = cleanupFrequency ?? TimeSpan.FromSeconds(10);
```
Resharper gives a minor concern, `Static Member in Generic Type` - do you permit these kinds of annotations in the source?
`// ReSharper disable once StaticMemberInGenericType`

StefH replied:
One way to solve this could be to have 2 classes: see example below....

``` c#
namespace WireMock.Util;

internal static class SingletonLock
{
    public static readonly object Lock = new();
}

internal static class SingletonFactory<T> where T : class, new()
{
    private static T? _instance;

    public static T GetInstance()
    {
        if (_instance == null)
        {
            lock (SingletonLock.Lock)
            {
                _instance ??= new T();
            }
        }

        return _instance;
    }
}
```

TWhidden replied:
Excellent idea - ugg, Sunday morning, I need a redbull!  Updated.

TWhidden commented on src/System.Linq.Dynamic.Core/Parser/ThreadSafeSlidingCache.cs:
```diff
@@ -0,0 +1,80 @@
+            var expirationTime = DateTime.UtcNow.Add(_timeToLive);
+            _cache[key] = (value, expirationTime);
+
+            CleanupIfNeeded();
```
Excellent idea. Great source reference. Will prevent the block.

TWhidden commented on src/System.Linq.Dynamic.Core/Parser/ConstantExpressionHelper.cs:
```diff
@@ -1,12 +1,18 @@
-        private static readonly ConcurrentDictionary<object, Expression> Expressions = new();
-        private static readonly ConcurrentDictionary<Expression, string> Literals = new();
+#if DEBUG
+        private static readonly TimeSpan TimeToLivePeriod = TimeSpan.FromSeconds(10);
```
I recommend keeping `ConstantExpressionHelper` as a static, non-configurable class. Its widespread use throughout the library suggests that refactoring it to manage a static instance elsewhere would involve significant changes. The current implementation seems to be the most straightforward and minimizes potential misuse.

Additionally, the simplicity of the Sliding Cache within `ConstantExpressionHelper` should be maintained. Overcomplicating it with configurations could lead to unnecessary complexity. The library itself handles the cleanup, which should remain transparent to the users. Just my .02c.

TWhidden commented on src/System.Linq.Dynamic.Core/Parser/ThreadSafeSlidingCache.cs:
```diff
@@ -0,0 +1,80 @@
+        private readonly TimeSpan _cleanupFrequency;
+        private DateTime _lastCleanupTime = DateTime.MinValue;
+
+        public ThreadSafeSlidingCache(TimeSpan timeToLive, TimeSpan? cleanupFrequency = null)
```
For the ConstantExpressionHelper, I believe that specifying the maximum number is not a critical concern for developers, given the unpredictability and variability of different projects. However, considering its potential utility as an aid to other parts of the library in the future, adding an optional configuration for this could be beneficial. Therefore, I have included it as an optional parameter in the constructor.

TWhidden commented on src/System.Linq.Dynamic.Core/Parser/ThreadSafeSlidingCache.cs:
```diff
@@ -0,0 +1,80 @@
+        private readonly TimeSpan _cleanupFrequency;
+        private DateTime _lastCleanupTime = DateTime.MinValue;
+
+        public ThreadSafeSlidingCache(TimeSpan timeToLive, TimeSpan? cleanupFrequency = null)
```
Great idea on the DateTime Mock. Implemented. I added `IDateTimeUtils` under Util, hopefully that is ok. (internal)

**Unit Testing:** New test cases for `ThreadSafeSlidingCache` have been added under the 'Util' folder in our test suite. These tests cover:
   - Adding, updating, removal operations.
   - `TryGetValue` functionality.
   - Expiry mechanism testing, including both manual expiry and automatic expiry scenarios (e.g., attempting to access a non-existent key and observing the auto-expiration of an existing key)."

TWhidden commented on src/System.Linq.Dynamic.Core/Parser/ThreadSafeSlidingCache.cs:
```diff
@@ -0,0 +1,83 @@
+
+        public void AddOrUpdate(TKey key, TValue value)
+        {
+            if (key == null) throw new ArgumentNullException(nameof(key));
```
Couldn't find usage of `Guard.NotNull` in the library, but did find `Check.NotNull(` - I assume that is what you wanted to use? Added a unit test for this also and throws correctly. `System.ArgumentNullException : Value cannot be null. (Parameter 'key')`

StefH replied:
> Couldn't find usage of `Guard.NotNull` in the library, but did find `Check.NotNull(` - I assume that is what you wanted to use? Added a unit test for this also and throws correctly. `System.ArgumentNullException : Value cannot be null. (Parameter 'key')`

**Check.NotNull** is the correct one indeed

TWhidden commented:
@StefH  - Ok, implemented the feedback above, with the exception of a configuration process as I think that really isn't needed for this (not sure who would be trying to tune this, this is just the library keeping itself clean so the GC can do its job).  

 - Added a Unit Test for the `ThreadSafeSlidingCache` and all its functions.
 - Removed the `ConstantExpressionHelper` since the unit tests for the `ThreadSafeSlidingCache` should be trusted. 
 - Dropped pre-processor directive I was using for tests, since implementing the Mock DateTime provider.

Thanks for the idea of the Mock DateTime - I can use that a lot on my own project tests!

StefH commented on test/System.Linq.Dynamic.Core.Tests/Util/ThreadSafeSlidingCacheTests.cs:
```diff
@@ -0,0 +1,152 @@
+            Assert.True(cache.Count == 1, $"Expected 1 items in the cache, had {cache.Count}");
+        }
+
+        private class MockDateTimeProvider : IDateTimeUtils
```
With this logic, you are still using a real DateTime.

The correct way to do is, is to Mock it:
``` c#
private static readonly DateTime UtcNow = new(2023, 1, 14, 15, 16, 17);

var dateTimeUtilsMock = new Mock<IDateTimeUtils>();
dateTimeUtilsMock.SetupGet(d => d.UtcNow).Returns(UtcNow);
```

See this test class:
https://github.com/WireMock-Net/WireMock.Net/blob/13a3dfa9feef7322064145a82bfe4ede50ef8974/test/WireMock.Net.Tests/MappingBuilderTests.cs#L45

TWhidden replied:
Implementing - I don't have any experience with that library - what is the proper way I should advance the time in my tests?  

I did this - seems to work, but not sure if there is a cleaner way to do it.

```
var r = dateTimeUtilsMock.Object.UtcNow.AddMinutes(11);
dateTimeUtilsMock.SetupGet(d => d.UtcNow).Returns(r);
```

StefH commented on test/System.Linq.Dynamic.Core.Tests/Util/ThreadSafeSlidingCacheTests.cs:
```diff
@@ -0,0 +1,152 @@
+            Assert.True(cache.Count == 1, $"Expected 1 items in the cache, only had {cache.Count}");
+
+            // move the time forward
+            mockDateTime.UtcNow = mockDateTime.UtcNow.AddMinutes(11);
```
Use **SetupSequence**


Example
``` c#
    public class YourTestClass
    {
        public void YourTestMethod()
        {
            // Create a mock of IDateTimeUtils
            var mockDateTimeUtils = new Mock<IDateTimeUtils>();

            // Set up a sequence for UtcNow to return different values
            mockDateTimeUtils.SetupSequence(x => x.UtcNow)
                             .Returns(new DateTime(2022, 1, 1))
                             .Returns(new DateTime(2022, 1, 2))
                             .Returns(new DateTime(2022, 1, 3));

            // Example usage
            Console.WriteLine(mockDateTimeUtils.Object.UtcNow); // Outputs 2022-01-01
            Console.WriteLine(mockDateTimeUtils.Object.UtcNow); // Outputs 2022-01-02
            Console.WriteLine(mockDateTimeUtils.Object.UtcNow); // Outputs 2022-01-03
        }
    }

```

TWhidden replied:
Oh sorry - missed this - disregard my last question. This makes more sense.

TWhidden replied:
I have a question regarding the cache class provider's frequent internal usage. To simplify the process and avoid the complexity of tracking the sequence of internal accesses, would it be more efficient to directly set the new time as needed? I'm concerned that the test class might become overly complicated if the cache is accessed multiple times internally before advancing the time. There are numerous internal comparisons, and detailing these in the code seems cumbersome. For my purposes, would it be feasible to implement a direct time-setting approach?

```
var r = dateTimeUtilsMock.Object.UtcNow.AddMinutes(11);
dateTimeUtilsMock.SetupGet(d => d.UtcNow).Returns(r);
```

StefH commented on test/System.Linq.Dynamic.Core.Tests/Util/ThreadSafeSlidingCacheTests.cs:
```diff
@@ -0,0 +1,152 @@
+            // Replace
+            cache.AddOrUpdate(1, "oneone");
+
+            Assert.True(cache.Count == 3, $"Expected 3 items in the cache, only had {cache.Count}");
```
Do not use `Assert.True` --> that is only used to check for a true boolean value.

If you want to check a normal value, use:
``` c#
Assert.Equal(3, cache.Count, $"Expected 3 items in the cache, only had {cache.Count}");
```

Or my personal preference : FluentAssertions:
``` c#
cache.Count.Should().Be(3, $"Expected 3 items in the cache, only had {cache.Count}");
```

TWhidden replied:
Having trouble with `Assert.Equal` when including a description. 
the `FluentAssertions` is quite slick though, going with that.

StefH commented on test/System.Linq.Dynamic.Core.Tests/Util/ThreadSafeSlidingCacheTests.cs:
```diff
@@ -0,0 +1,152 @@
+            mockDateTime.UtcNow = mockDateTime.UtcNow.AddMinutes(11);
+            if (cache.TryGetValue(1, out var value))
+            {
+                Assert.True(false, $"Expected to not find the value, but found {value}");
```
In this case, use `Asser.Fail`

TWhidden replied:
Lol, that's a Co-pilot fail!  Fixed! How did I miss that!

TWhidden replied:
So, going back to this - having trouble with xUnit - does not appear they have `Assert.Fail(..)` - I think Co-pilot suggested that as an alternative.  Some StackOverflow articles show I should just throw an exception instead, which will fail it. I am not sure how I feel about that.  How would you want me to trigger a failure?

StefH replied:
I understand. I was using an older version from xunit in the solution.
It's added here: https://github.com/xunit/xunit/issues/2609#issuecomment-1399639884

If you want; you upgrade all unit-test projects to use 2.5.0 which includes Assert.Fail

Else if this too much work, just keep the existing code.

StefH commented on src/System.Linq.Dynamic.Core/Util/ThreadSafeSlidingCache.cs:
```diff
@@ -0,0 +1,135 @@
+        ///     Cleanup the Cache
+        /// </summary>
+        /// <returns></returns>
+        private Task Cleanup()
```
Rename to CleanupAsync

StefH commented on src/System.Linq.Dynamic.Core/Util/ThreadSafeSlidingCache.cs:
```diff
@@ -0,0 +1,135 @@
+            }
+        }
+
+        /// <summary>
```
private methods do not need xml-comment, expect when very special or complex

StefH commented on src/System.Linq.Dynamic.Core/Util/ThreadSafeSlidingCache.cs:
```diff
@@ -0,0 +1,135 @@
+        private readonly ConcurrentDictionary<TKey, (TValue Value, DateTime ExpirationTime)> _cache;
+        private readonly TimeSpan _cleanupFrequency;
+        private readonly IDateTimeUtils _dateTimeProvider;
+        private readonly Func<Task> _deleteExpiredCachedItemsDelegate;
```
In this case, I think that this can be an `Action` (so the method is just a simple void method).

TWhidden replied:
In this case, will not add the `Async` to the end, since it's not async any longer.  Agree, this is probably better.

StefH commented on src/System.Linq.Dynamic.Core/Util/ThreadSafeSlidingCache.cs:
```diff
@@ -0,0 +1,135 @@
+
+        /// <summary>
+        ///     Check if cache needs to be cleaned up.
+        ///     If it does, span the cleanup as a Task to prevent from blocking
```
`If it does, start the cleanup as a Task to prevent from blocking`

StefH commented on src/System.Linq.Dynamic.Core/Util/ThreadSafeSlidingCache.cs:
```diff
@@ -0,0 +1,135 @@
+
+    internal class ThreadSafeSlidingCache<TKey, TValue> where TKey : notnull where TValue : notnull
+    {
+        private readonly ConcurrentDictionary<TKey, (TValue Value, DateTime ExpirationTime)> _cache;
```
"Named": Tuple is not supported in older frameworks like 4.5.x
![image](https://github.com/zzzprojects/System.Linq.Dynamic.Core/assets/249938/1f32e2a8-9e0b-47e7-8ecd-63bf7ae2e532)

So use `Tuple<TValue, DateTime>` or create a simple custom struct for this.

(Maybe create a sub-folder in "Utils" named "Cache", where you can place this file and also move ThreadSafeSlidingCacheConstants to an own class in that new folder.

TWhidden replied:
Probably would have picked all this up if I installed all the other frameworks on my dev system. Just hate having to install such legacy targets! I'll try and get that done tonight. I am not a huge fan of named tuples just yet, but it's very convenient for private items like this - but easy to swap, and kind of cleaner without.

StefH commented on src/System.Linq.Dynamic.Core/Util/ThreadSafeSlidingCache.cs:
```diff
@@ -0,0 +1,135 @@
+            CleanupIfNeeded();
+        }
+
+        public bool TryGetValue(TKey key, out TValue value)
```
Officially this Try should look like this example:

- `out TValue? value`
- `[NotNullWhen(true)] `

```` c#
public bool TryUnwrapAsValue<TValue>(Expression? expression, [NotNullWhen(true)] out TValue? value)
    {
        if (_parsingConfig.UseParameterizedNamesInDynamicQuery && _constantExpressionWrapper.TryUnwrapAsValue(expression as MemberExpression, out value))
        {
            return true;
        }

        value = default;   // <--------- just default, not default!
        return false;
    }
````

TWhidden replied:
Agree - a nice pattern especially when using nullable!

StefH commented on src/System.Linq.Dynamic.Core/Util/ThreadSafeSlidingCache.cs:
```diff
@@ -0,0 +1,135 @@
+                // Set here, so we don't have re-entry due to large collection enumeration.
+                _lastCleanupTime = _dateTimeProvider.UtcNow;
+
+                Task.Run(_deleteExpiredCachedItemsDelegate);
```
`Task.Run` does not exists for .NET 3.5

Make a new static class named `TaskUtils` and add a method `Run` which looks the same as Task.Run (with an `Action action`).

Add a preprocessor directive for .NET35 to use code like:
``` c#
ThreadPool.QueueUserWorkItem(_ => _deleteExpiredCachedItemsDelegate.Invoke());
```` 

And Else just use Task.Run

TWhidden commented on test/System.Linq.Dynamic.Core.Tests/Util/ThreadSafeSlidingCacheTests.cs:
```diff
@@ -0,0 +1,152 @@
+using Xunit;
+
+namespace System.Linq.Dynamic.Core.Tests.Util
+{
```
I **love** file-scoped namespaces, but the csproj `LangVersion` is not defined (or set to `Latest`). When I use Resharper to cleanup, it re-adds the scope brackets.  I manually set it for now, but may I suggest the `LangVersion` be defined in the `csproj` so things like the RS Code Cleanup will auto-refactor that?  

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version 

Not sure why RS doesn't honor C#12 by default for 8 projects, maybe because it has several Target Frameworks? not sure.  Let me know if I can add that in.

TWhidden commented on test/System.Linq.Dynamic.Core.Tests/Util/ConstantExpressionHelperTests.cs:
```diff
@@ -0,0 +1,48 @@
+﻿using System.Linq.Dynamic.Core.Parser;
```
I deleted it - Don't think it's needed any-longer with the new Test for the Sliding Cache we added. Ill add that to the next check-in.

TWhidden commented:
**Feedback on the Latest Check-In:**

1. **Singleton Implementation in `ConstantExpressionHelper`:** I've implemented an instance of `ConstantExpressionHelper` using a singleton pattern to prevent type duplication across instances. I'd appreciate any feedback on this pattern.

2. **Testing - Setup Sequence:** Seeking feedback on the setup sequence for tests. Setting up every datetime feels cumbersome, as the provider is extensively used within the Cache. Is it necessary to detail every access in the SetupSequence, or is setting the next time sufficient for these tests?

3. **Issues with .NET Framework and Project Configurations:**
   - Encountered issues with the .NET 4.0 targeting pack and .NET 4.5.2 compilation, unrelated to my changes. These issues stemmed from other types within the class.
   - I had to unload projects for .NET 3.5 (concerns about `TaskUtil`) and disable Android and UWP tests due to signing and certificate issues. My environment might require additional configuration for these tests.

4. **New ParserConfig Options:** Added three new `ParserConfig` options to reflect the different possibilities within the Cache of `ConstantExpressionHelper`, as suggested.

StefH commented on src/System.Linq.Dynamic.Core/Parser/ConstantExpressionHelper.cs:
```diff
@@ -1,29 +1,60 @@
     {
-        private static readonly ConcurrentDictionary<object, Expression> Expressions = new();
-        private static readonly ConcurrentDictionary<Expression, string> Literals = new();
+        _config = config;
```
`_config = Check.NotNull(config);`

StefH commented on src/System.Linq.Dynamic.Core/Parser/ConstantExpressionHelper.cs:
```diff
@@ -1,29 +1,60 @@
+        );
+        return _expressions;
+    }
+
```
newline

StefH commented on src/System.Linq.Dynamic.Core/Parser/ConstantExpressionHelper.cs:
```diff
@@ -1,29 +1,60 @@
+    private readonly ParsingConfig _config;
+
+    // Static shared instance to prevent duplications of the same objects
+    private static ThreadSafeSlidingCache<object, Expression>? _expressions;
```
In this case I think it's better to use a SingletonFactory pattern which takes care that the ConstantExpressionHelper is only created once.

Something like this:
``` c#
internal static class ConstantExpressionHelperFactory
{
    private static readonly object Lock = new();
    private static ConstantExpressionHelper? _instance;

    public static ConstantExpressionHelper GetInstance(ParsingConfig config)
    {
        if (_instance == null)
        {
            lock (Lock)
            {
                _instance ??= new ConstantExpressionHelper(config);
            }
        }

        return _instance;
    }
}
```

Based on this code:
https://github.com/WireMock-Net/WireMock.Net/blob/stef-protobuf/src/WireMock.Net/Util/SingletonFactory.cs

StefH commented on src/System.Linq.Dynamic.Core/Util/TaskUtils.cs:
```diff
@@ -0,0 +1,16 @@
+    {
+#if NET35 || NET40
+        System.Threading.ThreadPool.QueueUserWorkItem(_ => {
+            action?.Invoke();
```
1. This can be a one-liner:
```
 System.Threading.ThreadPool.QueueUserWorkItem(_ => action.Invoke());
```

2. No need to use `?` if you just add a Check.NotNull on line 6.

StefH commented on src/System.Linq.Dynamic.Core/Util/Cache/CacheContainer.cs:
```diff
@@ -0,0 +1,14 @@
+        ExpirationTime = expirationTime;
+    }
+
+    public TValue Value { get; }
```
Normally I put the properties and fields in the top of the file.

StefH reviewed (changes requested):
Thanks for the changes, only a few comments left. can you please take a look?

StefH commented on test/System.Linq.Dynamic.Core.Tests/Util/Cache/ThreadSafeSlidingCacheTests.cs:
```diff
@@ -0,0 +1,151 @@
+        // Act
+        cache.AddOrUpdate(1, "one");
+
+        var r = dateTimeUtilsMock.Object.UtcNow.AddMinutes(11);
```
please use a better name instead of `r`

StefH commented:
@TWhidden 
Also, if you have solved my comment, you can "Resolve" this so that comment is closed in this issue page. Else I get lost reading all the open comments (which some are actually resolved) ;-)

Like:
![image](https://github.com/zzzprojects/System.Linq.Dynamic.Core/assets/249938/b4d164f6-aaf4-4329-b484-db116e5ef910)

StefH commented on src/System.Linq.Dynamic.Core/ParsingConfig.cs:
```diff
@@ -234,4 +234,27 @@ public IQueryableAnalyzer QueryableAnalyzer
+    /// Default is 10 minutes.
+    /// </summary>
+    public TimeSpan ConstantExpressionSlidingCacheTimeToLive { get; set; } = TimeSpan.FromMinutes(10);
+
```
remove newline pls

StefH commented on src/System.Linq.Dynamic.Core/ParsingConfig.cs:
```diff
@@ -234,4 +234,27 @@ public IQueryableAnalyzer QueryableAnalyzer
+    /// A default value of null implies that cleanup is always allowed to run, helping in timely removal of unused cache items.
+    /// </summary>
+    public int? ConstantExpressionSlidingCacheMinItemsTrigger { get; set; } = null;
+
```
remove newline pls

StefH commented on src/System.Linq.Dynamic.Core/ParsingConfig.cs:
```diff
@@ -234,4 +234,27 @@ public IQueryableAnalyzer QueryableAnalyzer
+    /// Items not accessed within this TTL will be expired, allowing garbage collection to reclaim the memory.
+    /// Default is 10 minutes.
+    /// </summary>
+    public TimeSpan ConstantExpressionSlidingCacheTimeToLive { get; set; } = TimeSpan.FromMinutes(10);
```
In case we need a cache for other things, it's best to add a `CacheConfig` class with  3 values:
- TimeToLive
- MinItemsTrigger
- CleanupFrequency

And use this class here like:
``` c#
public CacheConfig? ConstantExpressionCacheConfig { get; set; }
```

TWhidden replied:
I've included a default `CacheConfig` instance instead of opting for null support. This approach allows developers to directly modify the existing cache settings, eliminating the need to create a new configuration instance.

`public CacheConfig ConstantExpressionCacheConfig { get; set; } = new CacheConfig();`

TWhidden commented:
My bad on "Resolve conversation" - that would have helped me also when going through them all.

Was thinking about the new `CacheConfig` - atm, it's just a POCO - should it be passed as an alternate constructor item to the `ThreadSafeSlidingCache` object? Just cleans up its usage a little bit, since the config is only for it.

Also - `ThreadSafeSlidingCache` just sounds silly - should we just call it `SlidingCache` ?  Later, if there is another kind of cache, it wont have that same naming pattern?

StefH commented:
> Also - `ThreadSafeSlidingCache` just sounds silly - should we just call it `SlidingCache` ? Later, if there is another kind of cache, it wont have that same naming pattern?

SlidingCache is also fine for me.

StefH commented:
> Was thinking about the new `CacheConfig` - atm, it's just a POCO - should it be passed as an alternate constructor item to the `ThreadSafeSlidingCache` object? Just cleans up its usage a little bit, since the config is only for it.

If it makes it easier, yes you can do it.

StefH commented:
@TWhidden 
Thanks for the PR, I'll merge it and see if it correctly compiles on my system.

TWhidden commented:
Hey @StefH  - We need to look a little into this.  I think with high contention, we are creating some errors.  Here is our Sentry report stack trace.

I propose some minor changes:
1.) in `CleanupIfNeeded`, use the `Interlocked.Exchange`, use this with a finally block as an atomic operation to keep possible multiple entries into Cleanup()
2.) `_cache[key]` I think is throwing this error; It may have been removed under contention, move to a `TryGet`
3.) In the `TryGetValue` of `SlidingCache` - instead of checking the expire time here, if it does exist, we should just update / return the value, Its a micro-optimization, but we are lazy removing anyways.  if something is requesting it, maybe we should allow it to be returned, or make it a configuration to permit expired returns;
![image](https://github.com/zzzprojects/System.Linq.Dynamic.Core/assets/9921122/3cd07ae9-f2a4-4637-8617-0a7c82ab4e9a)

If you are ok with it, I would love to do another PR for it. Hopefully with less hassle this time to your process.

TWhidden commented:
Ill Open a new issue - and submit a PR;
